File: examples/c/platform-failures/platform-failures.c
/* Copyright (c) 2007-2021. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

#include "simgrid/actor.h"
#include "simgrid/comm.h"
#include "simgrid/engine.h"
#include "simgrid/exec.h"
#include "simgrid/host.h"
#include "simgrid/mailbox.h"

#include "xbt/log.h"
#include "xbt/str.h"
#include "xbt/sysdep.h"

#include <stdio.h> /* snprintf */

#define FINALIZE 221297 /* a magic number to tell people to stop working */

XBT_LOG_NEW_DEFAULT_CATEGORY(platform_failures, "Messages specific for this example");

static void master(int argc, char* argv[])
{
  xbt_assert(argc == 5);
  long number_of_tasks  = xbt_str_parse_int(argv[1], "Invalid amount of tasks: %s");
  double task_comp_size = xbt_str_parse_double(argv[2], "Invalid computational size: %s");
  long task_comm_size   = xbt_str_parse_int(argv[3], "Invalid communication size: %s");
  long workers_count    = xbt_str_parse_int(argv[4], "Invalid amount of workers: %s");

  XBT_INFO("Got %ld workers and %ld tasks to process", workers_count, number_of_tasks);

  for (int i = 0; i < number_of_tasks; i++) {
    char mailbox_name[256];
    snprintf(mailbox_name, 255, "worker-%ld", i % workers_count);
    sg_mailbox_t mailbox = sg_mailbox_by_name(mailbox_name);

    XBT_INFO("Send a message to %s", mailbox_name);
    double* payload = (double*)xbt_malloc(sizeof(double));
    *payload        = task_comp_size;
    sg_comm_t comm  = sg_mailbox_put_async(mailbox, payload, task_comm_size);

    switch (sg_comm_wait_for(comm, 10.0)) {
      case SG_OK:
        XBT_INFO("Send to %s completed", mailbox_name);
        break;
      case SG_ERROR_NETWORK:
        XBT_INFO("Mmh. The communication with '%s' failed. Nevermind. Let's keep going!", mailbox_name);
        free(payload);
        break;
      case SG_ERROR_TIMEOUT:
        XBT_INFO("Mmh. Got timeouted while speaking to '%s'. Nevermind. Let's keep going!", mailbox_name);
        free(payload);
        break;
      default:
        xbt_die("Unexpected behavior");
    }
  }

  XBT_INFO("All tasks have been dispatched. Let's tell everybody the computation is over.");
  for (int i = 0; i < workers_count; i++) {
    char mailbox_name[256];
    snprintf(mailbox_name, 255, "worker-%ld", i % workers_count);
    sg_mailbox_t mailbox = sg_mailbox_by_name(mailbox_name);
    double* payload      = (double*)xbt_malloc(sizeof(double));
    *payload             = FINALIZE;
    sg_comm_t comm       = sg_mailbox_put_async(mailbox, payload, 0);

    switch (sg_comm_wait_for(comm, 1.0)) {
      case SG_ERROR_NETWORK:
        XBT_INFO("Mmh. Can't reach '%s'! Nevermind. Let's keep going!", mailbox_name);
        free(payload);
        break;
      case SG_ERROR_TIMEOUT:
        XBT_INFO("Mmh. Got timeouted while speaking to '%s'. Nevermind. Let's keep going!", mailbox_name);
        free(payload);
        break;
      case SG_OK:
        /* nothing */
        break;
      default:
        xbt_die("Unexpected behavior with '%s'", mailbox_name);
    }
  }

  XBT_INFO("Goodbye now!");
}

static void worker(int argc, char* argv[])
{
  xbt_assert(argc == 2);
  char mailbox_name[80];
  long id = xbt_str_parse_int(argv[1], "Invalid argument %s");

  snprintf(mailbox_name, 79, "worker-%ld", id);
  sg_mailbox_t mailbox = sg_mailbox_by_name(mailbox_name);

  while (1) {
    XBT_INFO("Waiting a message on %s", mailbox_name);
    double* payload;
    sg_comm_t comm     = sg_mailbox_get_async(mailbox, (void**)&payload);
    sg_error_t retcode = sg_comm_wait(comm);
    if (retcode == SG_OK) {
      if (*payload == FINALIZE) {
        free(payload);
        break;
      } else {
        double comp_size = *payload;
        free(payload);
        XBT_INFO("Start execution...");
        sg_actor_execute(comp_size);
        XBT_INFO("Execution complete.");
      }
    } else if (retcode == SG_ERROR_NETWORK) {
      XBT_INFO("Mmh. Something went wrong. Nevermind. Let's keep going!");
    }
  }
  XBT_INFO("I'm done. See you!");
}

int main(int argc, char* argv[])
{
  simgrid_init(&argc, argv);
  xbt_assert(argc > 2,
             "Usage: %s platform_file deployment_file\n"
             "\tExample: %s msg_platform.xml msg_deployment.xml\n",
             argv[0], argv[0]);

  simgrid_load_platform(argv[1]);

  simgrid_register_function("master", master);
  simgrid_register_function("worker", worker);
  simgrid_load_deployment(argv[2]);

  simgrid_run();

  XBT_INFO("Simulation time %g", simgrid_get_clock());

  return 0;
}

File: examples/python/platform-failures/platform-failures.py
## Copyright (c) 2007-2022. The SimGrid Team. All rights reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the license (GNU LGPL) which comes with this package.
 
from simgrid import Actor, Engine, Host, Mailbox, this_actor, NetworkFailureException, TimeoutException
import sys

# This example shows how to work with the state profile of a host or a link,
# specifying when the resource must be turned on or off.
#
# To set such a profile, the first way is to use a file in the XML, while the second is to use the programmatic
# interface, as exemplified in the main() below. Once this profile is in place, the resource will automatically
# be turned on and off.
#
# The actors running on a host that is turned off are forcefully killed
# once their on_exit callbacks are executed. They cannot avoid this fate.
# Since we specified on_failure="RESTART" for each actors in the XML file,
# they will be automatically restarted when the host starts again.
#
# Communications using failed links will .. fail.

def master(* args):
  assert len(args) == 4, f"Actor master requires 4 parameters, but got {len(args)} ones."
  tasks_count   = int(args[0])
  comp_size     = int(args[1])
  comm_size     = int(args[2])
  workers_count = int(args[3])

  this_actor.info(f"Got {workers_count} workers and {tasks_count} tasks to process")

  for i in range(tasks_count): # For each task to be executed: 
    # - Select a worker in a round-robin way
    mailbox = Mailbox.by_name(f"worker-{i % workers_count}")
    try:
      this_actor.info(f"Send a message to {mailbox.name}")
      mailbox.put(comp_size, comm_size, 10.0)
      this_actor.info(f"Send to {mailbox.name} completed")
    except TimeoutException:
      this_actor.info(f"Mmh. Got timeouted while speaking to '{mailbox.name}'. Nevermind. Let's keep going!")
    except NetworkFailureException:
      this_actor.info(f"Mmh. The communication with '{mailbox.name}' failed. Nevermind. Let's keep going!")

  this_actor.info("All tasks have been dispatched. Let's tell everybody the computation is over.")
  for i in range (workers_count):
    # - Eventually tell all the workers to stop by sending a "finalize" task
    mailbox = Mailbox.by_name(f"worker-{i % workers_count}")
    try:
      mailbox.put(-1.0, 0, 1.0)
    except TimeoutException:
      this_actor.info(f"Mmh. Got timeouted while speaking to '{mailbox.name}'. Nevermind. Let's keep going!")
    except NetworkFailureException:
      this_actor.info(f"Mmh. The communication with '{mailbox.name}' failed. Nevermind. Let's keep going!")

  this_actor.info("Goodbye now!")

def worker(* args):
  assert len(args) == 1, "Expecting one parameter"
  my_id               = int(args[0])

  mailbox = Mailbox.by_name(f"worker-{my_id}")
  done = False
  while not done:
    try:
      this_actor.info(f"Waiting a message on {mailbox.name}")
      compute_cost = mailbox.get()
      if compute_cost > 0: # If compute_cost is valid, execute a computation of that cost 
        this_actor.info("Start execution...")
        this_actor.execute(compute_cost)
        this_actor.info("Execution complete.")
      else: # Stop when receiving an invalid compute_cost
        this_actor.info("I'm done. See you!")
        done = True
    except NetworkFailureException:
      this_actor.info("Mmh. Something went wrong. Nevermind. Let's keep going!")

def sleeper():
  this_actor.info("Start sleeping...")
  this_actor.sleep_for(1)
  this_actor.info("done sleeping.")

if __name__ == '__main__':
  assert len(sys.argv) > 2, f"Usage: python app-masterworkers.py platform_file deployment_file"

  e = Engine(sys.argv)

  # This is how to attach a profile to an host that is created from the XML file.
  # This should be done before calling load_platform(), as the on_creation() event is fired when loading the platform.
  # You can never set a new profile to a resource that already have one.
  def on_creation(host):
    if (host.name == "Bourrassa"):
      host.set_state_profile("67 0\n70 1\n", 0)
  Host.on_creation_cb(on_creation)

  e.load_platform(sys.argv[1])

  e.register_actor("master", master)
  e.register_actor("worker", worker)
  e.load_deployment(sys.argv[2])

  # Add a new host programatically, and attach a state profile to it
  lili = e.netzone_root.create_host("Lilibeth", 1e15)
  lili.set_state_profile("4 0\n5 1\n", 10)
  lili.seal()

  # Create an actor on that new host, to monitor its own state
  actor = Actor.create("sleeper", lili, sleeper)
  actor.set_auto_restart(True)

  e.run()

  this_actor.info(f"Simulation time {e.clock():.4f}")
